### DISCUSSION TITLE
asyncio, socket and select

### DISCUSSION THREAD

**Discussion by fmafma** [micropython/micropython#15729] [General]
Hi!

In Peter's documentation about asyncio and socket:
https://github.com/peterhinch/micropython-async/blob/master/v3/docs/TUTORIAL.md#76-socket-programming
it is mentionned that there is an example using select in `client_server` dir:
https://github.com/peterhinch/micropython-async/blob/master/v3/as_drivers/client_server/userver.py
If the select module is imported, I don't see it used anywhere.

Did I miss something?

Thanks.

**Comment by sosi-deadeye** [comment]
The function `asyncio.start_server` binds the socket and if a new connection is made, the async function `run_client` is called with an instance of `StreamReader` and `StreamWriter`. If you await for data, the control is given back to the even loop. The even loop polls. I guess it uses select internally. In the second example, the import of select is not required.

**Comment by fmafma** [comment]
But it is written:

> The `userver` application uses `select.poll` **explicitly** to poll the server socket. The client sockets use it implicitly in that the `asyncio` stream mechanism employs it.

I would like to understand how to use `select.poll`, as I need to work at socket level, and not with streams...

**Comment by peterhinch** [comment]
Apologies. The tutorial was written for `uasyncio` V2, then updated for V3. The comment about `select` and the example import are outdated. I will update.

For examples that really use `select` see [this official example](https://github.com/micropython/micropython/blob/master/examples/network/https_client_nonblocking.py) and [this unofficial code](https://github.com/peterhinch/micropython-samples/blob/master/ntptime/ntptime.py). Also [this tutorial text](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/TUTORIAL.md#361-querying-a-threadsafeflag).

**Comment by fmafma** [comment]
Thank you for the links!

**Reply by peterhinch** [reply]
I've amended the tutorial text. Thanks for pointing out the error.
